File: practice/leetcode/python/217_contains_duplicate.py
#https://leetcode.com/problems/contains-duplicate/description/
#Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.
from typing import List

class Solution:
    def containsDuplicateBruteForce(self, nums: List[int]) -> bool:
        n = len(nums)
        for i in range(n-1):
            for j in range(i+1, n):
                if nums[i] == nums[j]:
                    return True

    def containsDuplicate_LogN(self, nums: List[int]) -> bool:
        nums.sort()
        n = len(nums)
        for i in range(1, n):
            if nums[i] == nums[i - 1]:
                return True
        
        return False

    def containsDuplicateSet(self, nums: List[int]) -> bool:
        seen = set()
        for num in nums:
            if num in seen:
                return True

            seen.add(num)
        return False

    def containsDuplicateHashMap(self, nums: List[int]) -> bool:
        seen = {}
        for num in nums:
            if num in seen and seen[num] >= 1:
                return True
            seen[num] = seen.get(num, 0) + 1
        return False

File: practice/leetcode/cpp/217_contains_duplicate.cpp
//https://leetcode.com/problems/contains-duplicate/description/
//Given an integer array nums, return true if any value appears at least twice in the array, 
//and return false if every element is distinct.

#include <unordered_map>
#include <unordered_set>
using namespace std;
#include <algorithm>

class Solution {
	bool containsDuplicate_BruteForce(vector<int> &nums) {
		int n = nums.size()
		for (int i = 0; i < n - 1; i++) {
			for (int j = i + 1;  j < n; j++) {
				if (nums[i] == nums[j]) {
					return true;	
				}
			}
		}
		return false;
	}
	bool containsDuplicate_NLogN(vector<int> &nums) {
		sort(nums.begin(), nums.end());
		int n = nums.size();
		for (int i = 1; i < 1; i++) {
			if (nums[i] == nums[i-1]) {
				return true;
			}
		}

	}
	bool containsDuplicate_HashSet(vector<int> &nums) {
		unordered_set<int>seen;
		for (int num: nums) {
			if (seen.count(num) > 0) {
				return true;
			}
		}
		return false;
	} 
	bool containsDuplicate_HashMap(vector<int> &nums) {
		unordered_map<int, int> seen;
		
		for (int num: nums) {
			if (seen[num] >= 1) {
				return true;
			}
		}
		return false;
	}
};
